File: linux/idsm_managerd/log_snapshot.cpp
/*
 * log_snapshot.cpp -- 日志快照分包上传(设计文档 11 章, 元消息 A.7)。
 * 持久目录布局:
 *   <dir>/pending/<transfer_id>/meta.json + chunk_NNNN(原始字节)
 *   <dir>/done/<transfer_id>/           全部发布完成(保留备查)
 *   <dir>/expired/<transfer_id>/        超过 24h 清理
 * meta.json: {transfer_id, filename, total_size, total_chunks, sha256,
 *   log_time, event_id, ecu_code, remark, created_ms, replay,
 *   meta_sent, next_chunks:[index...]}
 */
#include "log_snapshot.h"

#include "base64.h"

#include <nlohmann/json.hpp>

#include <openssl/evp.h>

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iterator>
#include <fcntl.h>
#include <random>
#include <sstream>
#include <unistd.h>

namespace idsm {

namespace fs = std::filesystem;

namespace {

std::string sha256HexFile(const fs::path& file, std::string& err) {
    std::ifstream is(file, std::ios::binary);
    if (!is) {
        err = "open " + file.string();
        return {};
    }
    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    if (!ctx) {
        err = "md ctx";
        return {};
    }
    EVP_DigestInit(ctx, EVP_sha256());
    char buf[64 * 1024];
    while (is) {
        is.read(buf, sizeof(buf));
        const auto n = is.gcount();
        if (n > 0) {
            EVP_DigestUpdate(ctx, buf, static_cast<size_t>(n));
        }
    }
    uint8_t digest[EVP_MAX_MD_SIZE];
    unsigned int dlen = 0;
    EVP_DigestFinal(ctx, digest, &dlen);
    EVP_MD_CTX_free(ctx);
    std::string out;
    out.reserve(dlen * 2);
    for (unsigned int i = 0; i < dlen; ++i) {
        char b[3];
        std::snprintf(b, sizeof(b), "%02x", digest[i]);
        out += b;
    }
    return out;
}

std::string sha256HexBytes(const uint8_t* data, size_t len) {
    uint8_t digest[EVP_MAX_MD_SIZE];
    unsigned int dlen = 0;
    EVP_Digest(data, len, digest, &dlen, EVP_sha256(), nullptr);
    std::string out;
    out.reserve(dlen * 2);
    for (unsigned int i = 0; i < dlen; ++i) {
        char b[3];
        std::snprintf(b, sizeof(b), "%02x", digest[i]);
        out += b;
    }
    return out;
}

std::string newTransferId() {
    std::random_device rd;
    const uint64_t a = (static_cast<uint64_t>(rd()) << 32) ^ rd();
    const uint64_t b = (static_cast<uint64_t>(rd()) << 32) ^ rd();
    char buf[33];
    std::snprintf(buf, sizeof(buf), "%016llx%016llx",
                  static_cast<unsigned long long>(a),
                  static_cast<unsigned long long>(b));
    return buf;
}

/* ISO8601 本地时间带时区偏移, 如 2026-09-18T10:20:30+08:00 */
std::string iso8601Local(long long ms) {
    const std::time_t t = static_cast<std::time_t>(ms / 1000);
    std::tm tmv {};
    localtime_r(&t, &tmv);
    char base[32];
    std::strftime(base, sizeof(base), "%Y-%m-%dT%H:%M:%S", &tmv);
    char zone[8];
    std::strftime(zone, sizeof(zone), "%z", &tmv);   /* +0800 */
    std::string z = zone;
    if (z.size() == 5) z.insert(3, ":");
    std::ostringstream os;
    os << base << "." << std::setw(3) << std::setfill('0')
       << static_cast<int>(ms % 1000) << z;
    return os.str();
}

bool saveMeta(const fs::path& dir, const nlohmann::json& meta,
              std::string& err) {
    const fs::path tmp = dir / "meta.json.tmp";
    {
        std::ofstream os(tmp, std::ios::binary | std::ios::trunc);
        os << meta.dump(1);
        os.flush();
        os.close();
        if (!os) {
            err = "meta write failed";
            return false;
        }
    }
    std::error_code ec;
    fs::rename(tmp, dir / "meta.json", ec);
    if (ec) {
        err = "meta rename: " + ec.message();
        return false;
    }
    return true;
}

nlohmann::json buildMetaEnvelope(const nlohmann::json& meta,
                                 const std::string& manufacturer,
                                 const std::string& ecu_code,
                                 long long now_ms) {
    nlohmann::json env = {
        {"msg_type", "log_snapshot"},
        {"protocol_version", "1.0"},
        {"timestamp", now_ms},
        {"manufacturer", manufacturer},
        {"transfer_id", meta.at("transfer_id")},
        {"filename", meta.at("filename")},
        {"total_size", meta.at("total_size")},
        {"total_chunks", meta.at("total_chunks")},
        {"sha256", meta.at("sha256")},
        {"log_time", meta.at("log_time")},
        {"ecuCode", ecu_code.empty()
                        ? meta.value("ecu_code", std::string()) : ecu_code},
        {"replay", meta.value("replay", false)},
    };
    if (meta.contains("event_id") && !meta["event_id"].get<std::string>().empty()) {
        env["event_id"] = meta["event_id"];
    }
    if (meta.contains("remark") && !meta["remark"].get<std::string>().empty()) {
        env["remark"] = meta["remark"];
    }
    return env;
}

}  /* namespace */

LogSnapshotManager::LogSnapshotManager(std::string snapshot_dir)
    : m_dir(std::move(snapshot_dir)) {}

std::string LogSnapshotManager::stageUpload(const SnapshotRequest& req,
                                            long long now_ms,
                                            std::string& err) {
    std::error_code ec;
    const fs::path file(req.file);
    if (!fs::is_regular_file(file, ec)) {
        err = "snapshot file not found: " + req.file;
        return {};
    }
    const auto size = fs::file_size(file, ec);
    if (ec || size == 0) {
        err = "snapshot file unreadable or empty: " + req.file;
        return {};
    }
    if (size > kSnapshotMaxFileSize) {
        err = "snapshot file too large (" + std::to_string(size) + " bytes)";
        return {};
    }

    const std::string sha = sha256HexFile(file, err);
    if (sha.empty()) return {};

    const std::string tid = newTransferId();
    const fs::path dir = fs::path(m_dir) / "pending" / tid;
    fs::create_directories(dir / "chunks", ec);
    if (ec) {
        err = "mkdir: " + ec.message();
        return {};
    }

    /* 切分片 */
    std::ifstream is(file, std::ios::binary);
    std::vector<char> buf(kSnapshotChunkSize);
    size_t total_chunks = 0;
    while (is) {
        is.read(buf.data(), static_cast<std::streamsize>(buf.size()));
        const auto n = is.gcount();
        if (n <= 0) break;
        char name[32];
        std::snprintf(name, sizeof(name), "chunk_%06zu", total_chunks);
        std::ofstream os(dir / "chunks" / name, std::ios::binary | std::ios::trunc);
        os.write(buf.data(), n);
        os.close();
        if (!os) {
            err = std::string("chunk write: ") + name;
            return {};
        }
        ++total_chunks;
    }

    nlohmann::json next = nlohmann::json::array();
    for (size_t i = 0; i < total_chunks; ++i) next.push_back(i);

    nlohmann::json meta = {
        {"transfer_id", tid},
        {"filename", file.filename().string()},
        {"total_size", size},
        {"total_chunks", total_chunks},
        {"sha256", sha},
        {"log_time", iso8601Local(now_ms)},
        {"event_id", req.event_id},
        {"ecu_code", std::string()},
        {"remark", req.remark},
        {"created_ms", now_ms},
        /* 断网缓存补传: 暂存超过 5 分钟才发布 -> replay=true(11.3) */
        {"replay", false},
        {"meta_sent", false},
        {"next_chunks", std::move(next)},
    };
    if (!saveMeta(dir, meta, err)) return {};
    return tid;
}

bool LogSnapshotManager::pump(const std::string& manufacturer,
                              const std::string& device_id,
                              const std::string& ecu_code,
                              long long now_ms,
                              const std::function<bool(const std::string&,
                                                       const std::string&)>& publish,
                              std::string& err) {
    const fs::path pending = fs::path(m_dir) / "pending";
    std::error_code ec;
    if (!fs::is_directory(pending, ec)) return false;

    /* 按创建时间排序, 先入先出 */
    std::vector<fs::path> transfers;
    for (const auto& e : fs::directory_iterator(pending, ec)) {
        if (e.is_directory()) transfers.push_back(e.path());
    }
    std::sort(transfers.begin(), transfers.end());

    bool progressed = false;
    for (const auto& dir : transfers) {
        nlohmann::json meta;
        {
            std::ifstream is(dir / "meta.json");
            if (!is) continue;
            try {
                is >> meta;
            } catch (...) {
                continue;
            }
        }

        /* 24h 过期清理(11.3) */
        if (now_ms - meta.value("created_ms", 0LL) > kSnapshotTransferTtlMs) {
            fs::create_directories(fs::path(m_dir) / "expired", ec);
            fs::rename(dir, fs::path(m_dir) / "expired" / dir.filename(), ec);
            std::fprintf(stderr, "[IDSMD] snapshot %s expired, moved to expired/\n",
                         dir.filename().c_str());
            continue;
        }

        const std::string topic = "oc/devices/" + device_id + "/sys/log/report";

        /* 元消息只发一次; 补传(replay)语义: 暂存超 5 分钟 */
        if (!meta.value("meta_sent", false)) {
            meta["replay"] = now_ms - meta.value("created_ms", now_ms) >
                             5LL * 60 * 1000;
            if (!publish(topic,
                         buildMetaEnvelope(meta, manufacturer, ecu_code,
                                           now_ms).dump())) {
                (void)saveMeta(dir, meta, err);
                return progressed;
            }
            meta["meta_sent"] = true;
            if (!saveMeta(dir, meta, err)) return progressed;
            progressed = true;
        }

        auto& next = meta["next_chunks"];
        int sent = 0;
        while (!next.empty() && sent < kSnapshotChunksPerPump) {
            const size_t idx = next[0].get<size_t>();
            char name[32];
            std::snprintf(name, sizeof(name), "chunk_%06zu", idx);
            std::ifstream cs(dir / "chunks" / name, std::ios::binary);
            if (!cs) {
                err = std::string("chunk missing: ") + name;
                return progressed;
            }
            std::string data((std::istreambuf_iterator<char>(cs)),
                             std::istreambuf_iterator<char>());
            const std::string chunk_sha =
                sha256HexBytes(reinterpret_cast<const uint8_t*>(data.data()),
                               data.size());
            const nlohmann::json env = {
                {"msg_type", "log_snapshot_chunk"},
                {"protocol_version", "1.0"},
                {"timestamp", now_ms},
                {"manufacturer", manufacturer},
                {"transfer_id", meta.at("transfer_id")},
                {"chunk_index", idx},
                {"chunk_sha256", chunk_sha},
                {"data", base64Encode(
                    reinterpret_cast<const uint8_t*>(data.data()), data.size())},
            };
            if (!publish(topic, env.dump())) {
                (void)saveMeta(dir, meta, err);
                return progressed;
            }
            next.erase(next.begin());
            ++sent;
            progressed = true;
        }

        if (next.empty()) {
            fs::create_directories(fs::path(m_dir) / "done", ec);
            fs::rename(dir, fs::path(m_dir) / "done" / dir.filename(), ec);
            std::fprintf(stderr, "[IDSMD] snapshot %s uploaded (%zu chunks)\n",
                         meta.value("transfer_id", std::string()).c_str(),
                         static_cast<size_t>(meta.value("total_chunks", 0ULL)));
        } else if (!saveMeta(dir, meta, err)) {
            return progressed;
        }
        if (sent >= kSnapshotChunksPerPump) return progressed;   /* 下轮继续 */
    }
    return progressed;
}

bool LogSnapshotManager::onNack(const std::string& payload, long long now_ms,
                                std::string& err) {
    nlohmann::json nack;
    try {
        nack = nlohmann::json::parse(payload);
    } catch (const std::exception& e) {
        err = std::string("nack parse: ") + e.what();
        return false;
    }
    const std::string tid = nack.value("transfer_id", std::string());
    if (tid.empty()) {
        err = "nack missing transfer_id";
        return false;
    }
    fs::path dir = fs::path(m_dir) / "pending" / tid;
    std::error_code ec;
    if (!fs::is_directory(dir, ec)) {
        /* 全部片已发布完毕(transfer 在 done/): 24h 窗口内云端仍可
         * 索补片(11.2), 移回 pending/ 重发缺失 index(幂等覆盖) */
        const fs::path done_dir = fs::path(m_dir) / "done" / tid;
        if (fs::is_directory(done_dir, ec)) {
            fs::rename(done_dir, dir, ec);
            if (ec) {
                err = "nack reactivate: " + ec.message();
                return false;
            }
        }
    }
    if (!fs::is_directory(dir, ec)) {
        err = "unknown or finished transfer: " + tid;
        return false;
    }
    nlohmann::json meta;
    {
        std::ifstream is(dir / "meta.json");
        if (!is) {
            err = "meta unreadable";
            return false;
        }
        try {
            is >> meta;
        } catch (const std::exception& e) {
            err = std::string("meta parse: ") + e.what();
            return false;
        }
    }
    /* 重发仅限 24 h 内(11.2) */
    if (now_ms - meta.value("created_ms", 0LL) > kSnapshotTransferTtlMs) {
        err = "transfer expired, nack ignored";
        return false;
    }
    if (!nack.contains("missing") || !nack["missing"].is_array() ||
        nack["missing"].empty()) {
        err = "nack missing indices";
        return false;
    }
    const size_t total = meta.value("total_chunks", 0ULL);
    std::vector<size_t> missing;
    for (const auto& v : nack["missing"]) {
        if (!v.is_number_unsigned() || v.get<size_t>() >= total) {
            err = "nack bad chunk index";
            return false;
        }
        missing.push_back(v.get<size_t>());
    }
    std::sort(missing.begin(), missing.end());
    missing.erase(std::unique(missing.begin(), missing.end()), missing.end());
    meta["next_chunks"] = missing;
    if (!saveMeta(dir, meta, err)) return false;
    std::fprintf(stderr, "[IDSMD] snapshot %s nack: %zu chunks to resend\n",
                 tid.c_str(), missing.size());
    return true;
}

}  /* namespace idsm */

File: android/idsm_manager/app/src/main/java/com/idsm/manager/LogSnapshot.kt
package com.idsm.manager

import android.content.Context
import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.security.SecureRandom
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * LogSnapshot -- 日志快照分包上传(设计 v1.0 第 11 章)。
 *
 * 与 linux/idsm_managerd/log_snapshot.cpp 同协议: >256KB 取证文件分包,
 * 元消息(A.7 扁平信封) + ≤128KB 分片(chunk_sha256 + base64 data),
 * 走 oc/devices/{device_id}/sys/log/report (QoS1)。
 *
 * 触发: 探针经 UDS 下行命令行 {"cmd":"log_upload","file":...,
 * "event_id":...,"remark":...}(与 managerd 一致, ProbeSocketServer 路由)。
 * 断网暂存 device-protected 目录 pending/<tid>/, 恢复后补传(replay=true,
 * 11.3); 云端整体 sha256 失败经 negative-ack 索补片, 24h 内重发(11.2)。
 */
class LogSnapshotManager(context: Context) {

    data class Request(
        val file: String,
        val eventId: String = "",
        val remark: String = "",
    )

    private val root: File = run {
        val dp = context.createDeviceProtectedStorageContext()
        File(dp.filesDir, "snapshots")
    }
    private val rng = SecureRandom()

    /** 暂存一次上传: 读文件、切分片、落盘 pending/<tid>/; 返回 transfer_id */
    fun stageUpload(req: Request, nowMs: Long): String? {
        val f = File(req.file)
        if (!f.isFile || f.length() == 0L || f.length() > MAX_FILE_SIZE) {
            Log.e(TAG, "snapshot rejected: ${req.file}")
            return null
        }
        val data = f.readBytes()
        val tid = newTransferId()
        val dir = File(root, "pending/$tid")
        val chunksDir = File(dir, "chunks")
        if (!chunksDir.mkdirs()) return null

        var count = 0
        var off = 0
        while (off < data.size) {
            val n = minOf(CHUNK_SIZE, data.size - off)
            File(chunksDir, chunkName(count)).writeBytes(data.copyOfRange(off, off + n))
            off += n
            ++count
        }
        val meta = JSONObject().apply {
            put("transfer_id", tid)
            put("filename", f.name)
            put("total_size", data.size)
            put("total_chunks", count)
            put("sha256", sha256Hex(data))
            put("log_time", isoLocal(nowMs))
            put("event_id", req.eventId)
            put("ecu_code", "")
            put("remark", req.remark)
            put("created_ms", nowMs)
            put("replay", false)
            put("meta_sent", false)
            put("next_chunks", JSONArray().apply {
                for (i in 0 until count) put(i)
            })
        }
        saveMeta(dir, meta)
        Log.i(TAG, "snapshot staged: $tid (${data.size}B x$count)")
        return tid
    }

    /** 发布进行中的 transfer; publish 返回 false 即停下轮续。无进展返回 false */
    fun pump(
        ecuCode: String,
        nowMs: Long,
        publish: (topic: String, payload: String) -> Boolean,
    ): Boolean {
        val pending = File(root, "pending")
        val dirs = pending.listFiles()?.filter { it.isDirectory }?.sortedBy { it.name }
            ?: return false
        var progressed = false
        for (dir in dirs) {
            val metaFile = File(dir, "meta.json")
            val meta = runCatching { JSONObject(metaFile.readText()) }.getOrNull()
                ?: continue

            /* 24h 过期清理(11.3) */
            if (nowMs - meta.optLong("created_ms") > TTL_MS) {
                File(root, "expired").mkdirs()
                dir.renameTo(File(root, "expired/${dir.name}"))
                continue
            }
            val topic = DeviceIdentity.snapshotTopic()

            if (!meta.optBoolean("meta_sent")) {
                /* 补传语义: 暂存超 5 分钟 -> replay=true(11.3) */
                meta.put("replay", nowMs - meta.optLong("created_ms") > 5 * 60 * 1000)
                val env = JSONObject().apply {
                    put("msg_type", "log_snapshot")
                    put("protocol_version", "1.0")
                    put("timestamp", nowMs)
                    put("manufacturer", DeviceIdentity.manufacturer)
                    put("transfer_id", meta.getString("transfer_id"))
                    put("filename", meta.getString("filename"))
                    put("total_size", meta.getInt("total_size"))
                    put("total_chunks", meta.getInt("total_chunks"))
                    put("sha256", meta.getString("sha256"))
                    put("log_time", meta.getString("log_time"))
                    put("ecuCode", ecuCode)
                    put("replay", meta.getBoolean("replay"))
                    if (meta.optString("event_id").isNotEmpty())
                        put("event_id", meta.getString("event_id"))
                    if (meta.optString("remark").isNotEmpty())
                        put("remark", meta.getString("remark"))
                }
                if (!publish(topic, env.toString())) {
                    saveMeta(dir, meta)
                    return progressed
                }
                meta.put("meta_sent", true)
                saveMeta(dir, meta)
                progressed = true
            }

            val next = meta.getJSONArray("next_chunks")
            var sent = 0
            while (next.length() > 0 && sent < CHUNKS_PER_PUMP) {
                val idx = next.getInt(0)
                val chunk = File(dir, "chunks/${chunkName(idx)}")
                if (!chunk.isFile) return progressed
                val raw = chunk.readBytes()
                val env = JSONObject().apply {
                    put("msg_type", "log_snapshot_chunk")
                    put("protocol_version", "1.0")
                    put("timestamp", nowMs)
                    put("manufacturer", DeviceIdentity.manufacturer)
                    put("transfer_id", meta.getString("transfer_id"))
                    put("chunk_index", idx)
                    put("chunk_sha256", sha256Hex(raw))
                    put("data", android.util.Base64.encodeToString(
                        raw, android.util.Base64.NO_WRAP))
                }
                if (!publish(topic, env.toString())) {
                    saveMeta(dir, meta)
                    return progressed
                }
                removeFirst(next)
                ++sent
                progressed = true
            }
            if (next.length() == 0) {
                File(root, "done").mkdirs()
                dir.renameTo(File(root, "done/${dir.name}"))
                Log.i(TAG, "snapshot uploaded: ${meta.getString("transfer_id")}")
            } else {
                saveMeta(dir, meta)
            }
            if (sent >= CHUNKS_PER_PUMP) return progressed
        }
        return progressed
    }

    /** negative-ack 消费端: 缺失 index 重排(24h 内, done/ 可重激活) */
    fun onNack(payload: String, nowMs: Long): Boolean {
        val nack = runCatching { JSONObject(payload) }.getOrNull() ?: return false
        val tid = nack.optString("transfer_id")
        if (tid.isEmpty()) return false
        var dir = File(root, "pending/$tid")
        if (!dir.isDirectory) {
            val done = File(root, "done/$tid")
            if (done.isDirectory) {
                dir = File(root, "pending/$tid")
                done.renameTo(dir)
            }
        }
        if (!dir.isDirectory) return false
        val metaFile = File(dir, "meta.json")
        val meta = runCatching { JSONObject(metaFile.readText()) }.getOrNull()
            ?: return false
        if (nowMs - meta.optLong("created_ms") > TTL_MS) return false
        val missing = runCatching { nack.getJSONArray("missing") }.getOrNull()
            ?: return false
        val total = meta.getInt("total_chunks")
        val indices = (0 until missing.length()).map { missing.getInt(it) }
            .filter { it in 0 until total }.sorted().distinct()
        if (indices.isEmpty()) return false
        meta.put("next_chunks", JSONArray().apply {
            indices.forEach { put(it) }
        })
        saveMeta(dir, meta)
        Log.i(TAG, "snapshot nack: $tid resend ${indices.size} chunks")
        return true
    }

    private fun newTransferId(): String {
        val b = ByteArray(16)
        rng.nextBytes(b)
        return b.joinToString("") { "%02x".format(it) }
    }

    private fun chunkName(idx: Int) = "chunk_%06d".format(idx)

    private fun saveMeta(dir: File, meta: JSONObject) {
        val tmp = File(dir, "meta.json.tmp")
        tmp.writeText(meta.toString(1))
        tmp.renameTo(File(dir, "meta.json"))
    }

    private fun removeFirst(arr: JSONArray) {
        /* org.json 无 remove(0) 以外的廉价头部删除, 整体重建 */
        val rest = JSONArray()
        for (i in 1 until arr.length()) rest.put(arr.get(i))
        for (i in 0 until arr.length()) arr.remove(0)
        for (i in 0 until rest.length()) arr.put(rest.get(i))
    }

    private fun sha256Hex(data: ByteArray): String =
        MessageDigest.getInstance("SHA-256").digest(data)
            .joinToString("") { "%02x".format(it) }

    private fun isoLocal(ms: Long): String =
        SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX", Locale.US)
            .format(Date(ms))

    companion object {
        private const val TAG = "IdsmSnapshot"
        const val CHUNK_SIZE = 128 * 1024      /* 每片原始字节上限(11.2) */
        const val TTL_MS = 24L * 3600 * 1000   /* transfer 有效期(11.3) */
        const val MAX_FILE_SIZE = 64L * 1024 * 1024
        const val CHUNKS_PER_PUMP = 8          /* 防饿死告警队列 */
    }
}
